RFC-0010-tensor-ref.md

```diff
@@ -0,0 +1,119 @@
+# TensorRef abstraction in C++
+
+This proposal introduces a new class `TensorRef`, to replace all
+places in our codebase where we currently use `const Tensor&`.  The
+distinguishing characteristics of this class are:
+
+1. It is non-owning
+2. It is as safe as other by-value reference types (like `c10::ArrayRef` or
+   `std::string_view`)
+3. It is implicitly convertible (with some exceptions) to `const Tensor&`
+   (i.e., it can be introduced incrementally)
+
+## Motivation
+
+* In order to call a function that accepts `const Tensor&`, you must
+  somehow have a `Tensor` at hand.  However, `Tensor` is always an
+  owning value type, and sometimes you only have a non-owning reference
+  to a Tensor, but not a reference to Tensor itself (e.g.,
+  `TensorImpl*`).  In these cases, you cannot conveniently call the
+  function without first inducing a refcount bump (to establish an
+  owning `Tensor` object, to pass by reference).
+
+* For performance reasons, IValue `toTensor` returns a `const Tensor&`.
+  For this to be possible, IValue must internally store a `Tensor`
+  object (otherwise, there is nothing to point the reference to.)  This
+  induces a dependency from IValue header to Tensor header, which can
+  cause problems in some situations (as Tensor class has a large API and
+  thus has many dependencies on many other classes.)
+
+* `const Tensor&` implies a double indirection, as `Tensor` is itself
+  already a pointer.  A TensorRef type implemented internally as a
+  `TensorImpl*` would eliminate this double indirection.  (NB: The
+  proposal below doesn't actually solve this problem, due to the definition
+  of Itanium ABI.)
+
+## Proposal
+
+Add the following class to PyTorch:
+
+```
+struct TensorRef {
+    Tensor base_;
+    TensorRef(TensorImpl* impl) : base_(TensorPtr::reclaim(impl)) {}
+    ~TensorRef() {
+        base_.impl_.release();
+    }
+    operator const Tensor& () & {
+        return base_;
+    }
+    // Generated copy of all Tensor methods, forwarding to base_
+    // (with a hypothetical C++ language extension for operator dot,
+    // would be possible to do this without codegen)
+};
+```
+
+Main points:
+
+* This class is implemented as containing an actual Tensor object, but
+  we unsafely use reclaim/release to ensure that no refcount bump occurs
+  on construction/destruction of the object.  We verified with Godbolt
+  that the compiler is able to eliminate the `base_` destructor.  This
+  makes it possible to take out a `const Tensor&`, which makes it easier
```

Comment:
as long as the `Tensor` is purely internal to `TensorRef`, this could be safe since we have control over the invariants and where refcounting occurs. But allowing users to get a `const Tensor&` out of it means that the "internal" Tensor object could be copied outside of your control and you have to be much more careful about correctness, especially since Tensor doesn't seem to know if it's borrowed or not. I saw one of your earlier proposals had Tensor know about this, maybe we should do that?

Or, if the `const Tensor&` conversion is only there to avoid having to rewrite all ops at once, maybe there's a regex/codemod that would allow us to rewrite them more easily and then not allow TensorRef -> const Tensor& conversions? Or, if we manage to eliminate most but not all call sites, we might be able to make it a TensorRef -> Tensor (by value) conversion and be ok with the refcount if somebody does that. Seems like a less dangerous design.

Reply:
> But allowing users to get a const Tensor& out of it means that the "internal" Tensor object could be copied outside of your control and you have to be much more careful about correctness

A copy of `const Tensor&` to `Tensor` is OK, because this always induces a refcount bump! Most of the bad situations @swolchok was able to come with required you to have a mutable reference/pointer to the internal Tensor, which we just forbid here.

> Or, if the const Tensor& conversion is only there to avoid having to rewrite all ops at once, maybe there's a regex/codemod that would allow us to rewrite them more easily and then not allow TensorRef -> const Tensor& conversions?

Yeah. I allude to this in the second alternative proposal, where for Intel ABI reasons, it is much better if the long term API doesn't permit conversion to `const Tensor&` (so we can make the class trivial). So ideally we'd get rid of this BC crutch eventually.